src/Features/CSharp/Portable/Formatting/CSharpFormattingInteractionService.cs

```diff
@@ -0,0 +1,463 @@
+            // work of attempting to read in the editorconfig file.
+            var options = document.Project.Solution.Workspace.Options;
+
+            var smartIndentOn = options.GetOption(FormattingOptions.SmartIndent, LanguageNames.CSharp) == FormattingOptions.IndentStyle.Smart;
+
+            // We consider the proper placement of a close curly or open curly when it is typed at
+            // the start of the line to be a smart-indentation operation.  As such, even if "format
+            // on typing" is off, if "smart indent" is on, we'll still format this.  (However, we
+            // won't touch anything else in the block this close curly belongs to.).
+            //
+            // See extended comment in GetFormattingChangesAsync for more details on this.
+            if (smartIndentOn)
+            {
+                if (ch == '{' || ch == '}')
+                {
+                    return true;
+                }
+            }
+
+            // If format-on-typing is not on, then we don't support formatting on any other characters.
+            var autoFormattingOnTyping = options.GetOption(FormattingOptions2.AutoFormattingOnTyping, LanguageNames.CSharp);
+            if (!autoFormattingOnTyping)
+            {
+                return false;
+            }
+
+            if (ch == '}' && !options.GetOption(BraceCompletionOptions.AutoFormattingOnCloseBrace, LanguageNames.CSharp))
+            {
+                return false;
+            }
+
+            if (ch == ';' && !options.GetOption(FormattingOptions2.AutoFormattingOnSemicolon, LanguageNames.CSharp))
+            {
+                return false;
+            }
+
+            // don't auto format after these keys if smart indenting is not on.
+            if ((ch == '#' || ch == 'n') && !smartIndentOn)
+            {
+                return false;
+            }
+
+            return _supportedChars.Contains(ch);
+        }
+
+        public async Task<IList<TextChange>> GetFormattingChangesAsync(
+            Document document,
+            TextSpan? textSpan,
+            DocumentOptionSet? documentOptions,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var span = textSpan ?? new TextSpan(0, root.FullSpan.Length);
+            var formattingSpan = CommonFormattingHelpers.GetFormattingSpan(root, span);
+            var options = documentOptions;
+            if (options == null)
+            {
+                var inferredIndentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IInferredIndentationService>();
+                options = await inferredIndentationService.GetDocumentOptionsWithInferredIndentationAsync(document, explicitFormat: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
```

Comment:
This was changed from a simple null-coalescing to this `if`, to extract out getting the `inferredIndentationService`.